website/addons/onedrive/model.py

```diff
@@ -0,0 +1,240 @@
+    folder_id = fields.StringField(default=None)
+    onedrive_id = fields.StringField(default=None)
+    folder_name = fields.StringField()
+    folder_path = fields.StringField()
+
+    _folder_data = None
+
+    _api = None
+
+    @property
+    def api(self):
+        """authenticated ExternalProvider instance"""
+        if self._api is None:
+            self._api = OneDrive(self.external_account)
+        return self._api
+
+    @property
+    def display_name(self):
+        return '{0}: {1}'.format(self.config.full_name, self.folder_name)
+
+    @property
+    def has_auth(self):
+        """Whether an access token is associated with this node."""
+        return bool(self.user_settings and self.user_settings.has_auth)
+
+    @property
+    def complete(self):
+        return bool(self.has_auth and self.user_settings.verify_oauth_access(
+            node=self.owner,
+            external_account=self.external_account,
+        ))
+
+    def fetch_folder_name(self):
+        self._update_folder_data()
+        return self.folder_name.replace('All Files', '/ (Full OneDrive)')
+
+    def fetch_full_folder_path(self):
+        self._update_folder_data()
+        return self.folder_path
+
+    def _update_folder_data(self):
+        if self.folder_id is None:
+            return None
+
+        logger.debug('self::' + repr(self))
+        #request.json.get('selected')
+
+        if not self._folder_data:
+            self.path = self.folder_name
+            self.save()
+
+    def set_folder(self, folder, auth):
+        self.folder_id = folder['name']
+        self.onedrive_id = folder['id']
+        self.folder_name = folder['name']
+        self.save()
+
+        if not self.complete:
+            self.user_settings.grant_oauth_access(
+                node=self.owner,
+                external_account=self.external_account,
+                metadata={'folder': self.folder_id}
+            )
+            self.user_settings.save()
+
+        # Add log to node
+        nodelogger = OneDriveNodeLogger(node=self.owner, auth=auth)  # AddonOAuthNodeSettingsBase.nodelogger(self)
+        nodelogger.log(action="folder_selected", save=True)
+
+    def set_user_auth(self, user_settings):
+        """Import a user's OneDrive authentication and create a NodeLog.
+
+        :param OneDriveUserSettings user_settings: The user settings to link.
+        """
+        self.user_settings = user_settings
+        nodelogger = OneDriveNodeLogger(node=self.owner, auth=Auth(user_settings.owner))
+        nodelogger.log(action="node_authorized", save=True)
+
+    def deauthorize(self, auth=None, add_log=True):
+        """Remove user authorization from this node and log the event."""
+        node = self.owner
+
+        if add_log:
+            extra = {'folder_id': self.folder_id}
+            nodelogger = OneDriveNodeLogger(node=node, auth=auth)
+            nodelogger.log(action="node_deauthorized", extra=extra, save=True)
+
+        self.folder_id = None
```

Comment:
Please write a `def clear_settings(self)` that sets this and `folder_name` to `None`. It's expected to exist by the base classes.